common/src/main/java/com/microsoft/identity/common/internal/controllers/ExceptionAdapter.java

```diff
@@ -137,111 +130,131 @@ public static BaseException exceptionFromAcquireTokenResult(final AcquireTokenRe
+        // We simplify our logic because this layer is also used by MSAL.
+        if (oAuthError.equalsIgnoreCase(AuthenticationConstants.OAuth2ErrorCode.INVALID_GRANT)){
+            return true;
         }
 
-        return null;
+        return false;
+    }
+
+    /**
+     * Get an exception object from the given oAuth values.
+     *
+     * @param oAuthError
+     * @param oAuthSubError
+     * @param oAuthErrorDescription
+     * @param response
+     * @return ServiceException, UiRequiredException
+     * */
+    public static ServiceException getExceptionFromOAuthError(final String oAuthError,
+                                                              final String oAuthSubError,
+                                                              final String oAuthErrorDescription,
+                                                              final HttpResponse response) {
+        final String methodName = ":getExceptionFromOAuthError";
+
+        ServiceException outErr;
+
+        if (shouldBeConvertedToUiRequiredException(oAuthError, oAuthSubError)) {
+            outErr = new UiRequiredException(
+                    oAuthError,
+                    oAuthErrorDescription);
+        }
+        else {
+            outErr = new ServiceException(
+                    oAuthError,
+                    oAuthErrorDescription,
+                    null);
+        }
+
+        outErr.setSubErrorCode(oAuthSubError);
+
+        try {
+            outErr.setHttpResponse(response);
+        }
+        catch (JSONException e) {
+            Logger.warn(
+                    TAG + methodName,
+                    "Failed to deserialize error data: status, headers, response body."
+            );
+        }
+
+        return outErr;
     }
 
-    private static void applyCliTelemInfo(@NonNull final TokenResult tokenResult,
+    public static void applyCliTelemInfo(@NonNull final CliTelemInfo cliTelemInfo,
                                           @NonNull final BaseException outErr) {
-        if (null != tokenResult.getCliTelemInfo()) {
-            final CliTelemInfo cliTelemInfo = tokenResult.getCliTelemInfo();
+        if (null != cliTelemInfo) {
             outErr.setSpeRing(cliTelemInfo.getSpeRing());
             outErr.setRefreshTokenAge(cliTelemInfo.getRefreshTokenAge());
             outErr.setCliTelemErrorCode(cliTelemInfo.getServerErrorCode());
             outErr.setCliTelemSubErrorCode(cliTelemInfo.getServerSubErrorCode());
         }
     }
 
-    private static void applyHttpErrorResponseData(@NonNull final ServiceException targetException,
-                                                   final int statusCode,
-                                                   @Nullable String responseHeadersJson,
-                                                   @Nullable String responseBody) {
+    private static HttpResponse synthesizeHttpResponse(final int statusCode,
```

Comment:
What's consuming HTTP Response above that could not consume the TokenErrorResponse?

Reply:
None. It's just that getExceptionFromOAuthError is also consumed in PRTController, and over there it parses httpResponse along with the error codes. The only TokenErrorResponse have are strings, so we have to convert them back.